### ISSUE TITLE
[java] UseConcurrentHashMap should consider non-thread safe type in return value

### ISSUE THREAD

**Issue by MetaBF** [pmd/pmd#3865]
**Affects PMD Version: 6.43.0** 

**Rule: UseConcurrentHashMap**
 https://pmd.github.io/latest/pmd_rules_java_multithreading.html#useconcurrenthashmap

**Description:**
Hi, I found a false-negative about rule **UseConcurrentHashMap**. In the following example, function `getMap()` returns a deep-copy of `threadSafeMap` with type `HashMap`, which is NOT thread-safe. Therefore, PMD should report a violation at line 6.

**Code Sample demonstrating the issue:**

``` java
public class Concurrent {

    private Map<String, String> threadSafeMap = new ConcurrentHashMap<>();

    public Map<String, String> getMap() {
        return new HashMap<>(threadSafeMap); // line 6: false-negative
    }
}
```

**Expected outcome:**

PMD should report a violation at line 6, but doesn't. This is a false-negative.

**Running PMD through:** *[CLI]*

**Comment by oowekyala** [comment]
I'm not sure this is a false-negative... Doing a copy of a thread-safe map into a non-thread safe map is a common pattern to process a snapshot of the map within a single method for instance. I think this code could be refactored in several ways:
- you could encapsulate the map more strongly, ie not expose a getter but other more high-level methods on the `Concurrent` class. This is IMO necessary if, like here, you don't want to advertise the thread-safety of the map 
- if you want to advertise it, you should probably reflect that by using the interface `ConcurrentMap` instead of `Map`
- you're doing a copy, as if not to allow the caller to mutate the internal map. You could also be returning a `Collections.unmodifiableMap`. If the caller needs mutability, they can choose to clone the map into either a `ConcurrentMap` or a regular `HashMap`, depending on how they use the map.

I think the last approach is the most flexible tbh. It makes no sense to have a getter return a `new ConcurrentHashMap<>(threadSafeMap)`, as the caller might not need thread-safety or mutability at all. With this approach, there is also no copy of the map unless the caller wants mutability, which is nice for performance.

It's not clear to me otherwise how to report this without also reporting a lot of valid code, as I think there is no fundamental problem in the code you show.

**Comment by MetaBF** [comment]
I appreciate your time, `Collections.unmodifiableMap` is indeed a better alternative for this case. 

While I still have some confusion about this rule :
* Now this rule supports the detection of the named local variables, then I am confused about why this rule missed anonymous variables. Please note that these anonymous variables can exist in not only return but also arguments (e.g., `func(new HashMap<>())`).
* It looks like this rule cannot detect declare and use in separated statements, like: 

``` java
public class Example {

    Map<String,String> hashMap; 

    Example(int initialCap) {
        hashMap = new HashMap<>(initialCap); 
    } 
}

```

**Comment by oowekyala** [comment]
You're right, the rule only cares about variable declarations with a very specific form. I had noticed this and actually proposed to replace the rule instead of extending it to these cases (https://github.com/pmd/pmd/issues/2926), because flagging any `new HashMap` is just not something we should do. Assuming #2926 is implemented, these FNs would be ignored as "won't fix"
